### PR TITLE
Dialcrowd

### PR DESCRIPTION
**Pull request by jessicah25**
**Patch description**
DialCrowd is a dialogue crowdsourcing toolkit that helps requesters write clear HITs, view and analyze results, and obtain higher-quality data. This integration allows for the requester interface, worker interface, and analysis interface to be integrated into ParlAI so requesters can have access to ParlAI's tools with DialCrowd's tools.

**Testing steps**
The testing steps can be found in the README; there are 3 components: configuration page, annotation page, quality page. The configuration script should allow a download of a config.json file, to be manually put into the task_config folder. The annotation page should use the config.json file to load a HIT that a worker can fully do. The quality page should use the information pulled from Mephisto for the annotation page to display the results and analysis.

### CODE REVIEW

**Review by JackUrb — COMMENTED** [review]
Hi @jessicah25, thanks for putting this forward! It's exciting to see what seems like a significant quality of life and ease of use improvement for dialogue tasks on Mephisto. At the moment though this PR is pretty monolithic and hard to parse. Particularly I'm not sure I follow what the use of the `webapp-config` and `webapp-results` are specifically, and can't tell if they're attempting to integrate with `mephisto`'s existing tooling for building out custom review flows or not (if not, I'm happy to help integrate!). It would be great to see some screenshots that break down your usage a little more clearly.

There are also a number of classes that would need to be renamed (primarily contents in your `parlai/crowdsourcing/tasks/dialcrowd/dialcrowd_blueprint.py` file) as they're overwriting the namespace of classes we use elsewhere. 

I can give a more detailed review on the whole PR after I get some more context back here.

File: parlai/crowdsourcing/tasks/dialcrowd/README.md
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/dialcrowd/README.md:1]: Would be very useful to include some information on what DialCrowd does, high-level aims, etc. here

Comment (jessicah25) [reply parlai/crowdsourcing/tasks/dialcrowd/README.md:1]: Hi! Thank you for the response; we do have a paper describing a previous version of DialCrowd available (https://aclanthology.org/W18-5028/), but a paper describing this version is yet to be available.

For the configuration page, this is to set up config.json, which populates the annotation page. We have sections which walk a requester through all the steps necessary to set up a HIT with this UI.
ss11

For the results page, this is to view the results afterward. It does some basic analysis, including checking if the golden data responses match, if a worker took an abnormally long or short time on the HIT, etc.
ss12

We also have a site, https://cmu-dialcrowd.herokuapp.com/, of which the intent classification tab has approximately the same functionalities as the template we have integrated into the ParlAI system.

Do let me know if there is an easier way to do any of these pages (and what other information you would want)!

**Review by JackUrb — COMMENTED** [review]
I've started to take a pass through the code, but I'm having difficulty disambiguating between required content and old stuff that exists because you were building off of some of our directories as a template. There are some interesting things here in `DialCrowd` that I really feel like ParlAI users would benefit by having, but it would be hard to maintain this codebase in the current state. I've left some early comments, but I'm going to hold off on a complete pass until things are cleaned up a bit.

File: parlai/crowdsourcing/tasks/dialcrowd/analysis/compile_results.py
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/dialcrowd/analysis/compile_results.py:1]: Yeah, files like this seem to be largely copied from current code. Can these be removed from the PR or are they used in some way? If they're being used, it'd be good to remove the duplication with the existing code

File: parlai/crowdsourcing/tasks/dialcrowd/run.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/dialcrowd/run.py:18]: > @jessicah25 I'm assuming you meant to import from your version of this util, as included in this PR?
> 
> @EricMichaelSmith as an aside, I imagine if this util is usable outside of just `turn_annotations_static` it could be good to pull out in a general form, but I'm not positive that this is perfectly generalizable even here.

Good point, making a note to look into this

Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/dialcrowd/run.py:18]: > @jessicah25 I'm assuming you meant to import from your version of this util, as included in this PR?
> 
> @EricMichaelSmith as an aside, I imagine if this util is usable outside of just `turn_annotations_static` it could be good to pull out in a general form, but I'm not positive that this is perfectly generalizable even here.

Good point, making a note to look into this

File: parlai/crowdsourcing/tasks/dialcrowd/run_in_flight_qa.py
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/dialcrowd/run_in_flight_qa.py:23]: @JackUrb Yes, @meganung recently merged the functionality of `turn_annotations_static/run_in_flight_qa.py` into `run.py` as part of her cleanup work from the Hydra 1.1 upgrade :)

**Review by EricMichaelSmith — COMMENTED** [review]
Just a few comments, some addressing @JackUrb's questions. But yeah, agreed with Jack, happy to take a deeper pass of this once we get a sense of whether there's a way to avoid this level of duplication of existing code from `turn_annotations_static/` :)

**Comment by jessicah25** [comment]
Hi! Sorry for the delay; but I've cleaned up the code so it hopefully is more clear, and removed a lot of the unneeded materials. Not sure how to reopen this since it got closed.

**Comment by JackUrb** [comment]
The cleanup changes here are hugely appreciated from my perspective, thanks for doing so @jessicah25! I still have a few unresolved questions above (mostly just to better understand the functionality), but overall one last thing that would be greatly appreciated would be including the screenshots from walking through an example into your readme - I find this helps set expectations and (short of having automated testing) makes it easier to know when something isn't quite working the way it was intended.

File: parlai/crowdsourcing/tasks/dialcrowd/README.md
Comment (EricMichaelSmith) [reply parlai/crowdsourcing/tasks/dialcrowd/README.md:1]: oh yeah thanks for this detail! Hmm would you be able to add this info to the README itself so that people can see it first thing when they view the README? It'd also be useful to add in your info in the PR description, "DialCrowd is a dialogue crowdsourcing toolkit that helps requesters write clear HITs, view and analyze results, and obtain higher-quality data. This integration allows for the requester interface, worker interface, and analysis interface to be integrated into ParlAI so requesters can have access to ParlAI's tools with DialCrowd's tools."

File: parlai/crowdsourcing/tasks/dialcrowd/README.md
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/dialcrowd/README.md:82]: hmm I think the blueprint has since been renamed?

File: parlai/crowdsourcing/tasks/dialcrowd/run.py
Comment (EricMichaelSmith) [inline parlai/crowdsourcing/tasks/dialcrowd/run.py:25]: nit: we've been centralizing Hydra defaults in the yaml files as opposed to the run.py files for consistency. See https://github.com/facebookresearch/ParlAI/blob/main/parlai/crowdsourcing/tasks/acute_eval/hydra_configs/conf/base_fast_acute.yaml for instance

**Review by EricMichaelSmith — COMMENTED** [review]
Oh, much cleaner, thanks so much for this :) A few more comments, but I think this is a massive step forward already

@JackUrb historically we've always had unit test coverage on everything in `parlai/crowdsourcing/tasks/`, but I think you said that maybe this isn't as crucial for now because a better solution for Mephisto E2E task testing its on its way? (Or do you think that it's still worth getting tests on this now?)

**Comment by JackUrb** [comment]
> a better solution for Mephisto E2E task testing its on its way

Definitely! Though we'd be hard pressed to know exactly what the baseline expected inputs and outputs are without a provided sample. Would definitely appreciate a minimal runtime inputs+outputs included in some `/tests/` folder, such that we can extend to automated testing in the future.

**Comment by jessicah25** [comment]
We were planning to link to our paper, which will be published soon, which has more information about DialCrowd. I can definitely add more information into the README. To address the questions you had above @JackUrb , DialCrowd is split into 3 components which are run separately: the configuration page, the MTurk worker page, and the results page. When we save the configuration file from the configuration page, it shouldn't be an issue of where the file is saved to since it should be on whatever machine the user is accessing ParlAI from. The configuration page and the results page aren't seen by the workers and are only served locally.

**Comment by jessicah25** [comment]
What kinds of tests would you want to include?

**Comment by EricMichaelSmith** [comment]
> What kinds of tests would you want to include?

Hmm take a look at the tests in `tests/crowdsourcing/tasks/`, for instance `tests/crowdsourcing/tasks/turn_annotations_static/test_turn_annotations_static.py` - this is what we typically do. It's a test for the backend code to make sure that the right agent state is produced given sample dummy inputs (as mentioned by @JackUrb )

**Comment by jessicah25** [comment]
I've made the other edits, but I'm working on the tests now, and I had a few questions. Would this test be applicable to the two sections of DialCrowd which are not worker-facing?

**Comment by jessicah25** [comment]
Just committed my changes! Let me know if I interpreted the tests correctly.

**Comment by EricMichaelSmith** [comment]
Cool, thanks a lot @jessicah25 for adding lots of user-friendly task description in the README. The test looks great! Yeah, this was what I was thinking of for a test, although maybe @JackUrb has other ideas as well.

Hmm it looks like a bunch of CI checks are broken, but it's unclear which of them are because of the changes in this PR, since this fork was split off from the `facebookresearch:main` branch a few months ago. Would you be able to merge from or rebase onto the HEAD of `facebookresearch:main`" in order to see if the CI checks pass given the latest code in that branch?

**Comment by jessicah25** [comment]
Yes, let me try to do that now!

**Comment by EricMichaelSmith** [comment]
> Yes, let me try to do that now!

Okay great! Hmm, looks like the unittests_38 CI check is failing now with the following error:
```
E           AssertionError: '__init__.py' not found in ['annotation1.png', 'annotation2.png', 'config1.png', 'config2.png', 'config3.png', 'config4.png', 'config5.png', 'config6.png', 'results1.png', 'results2.png', 'results3.png'] : parlai/crowdsourcing/tasks/dialcrowd/images does not contain __init__.py
```
However, this should be an easy fix (there just needs to be a `__init__.py` file added to that folder)

**Comment by jessicah25** [comment]
It seems like there are other errors, but I'm not sure what is causing them; would you be able to give some insight into it?

**Comment by jessicah25** [comment]
Looks like the remaining errors are the ones still existing on main (long_gpu_tests, teacher_tests, unittests_gpu18), and crowdsourcing_tests!

**Comment by EricMichaelSmith** [comment]
> Looks like the remaining errors are the ones still existing on main (long_gpu_tests, teacher_tests, unittests_gpu18), and crowdsourcing_tests!

Hmm it looks like most of the linting issues are resolved, but there's still one left (apparently I have to approve in order to run the linting CI check): the error message says it's on line 76 of `tests/crowdsourcing/tasks/dialcrowd/test_dialcrowd.py`. Can you do `pip install black` and run it on that script to try to fix that error? Black is sometimes very picky about minor linting issues

**Comment by jessicah25** [comment]
Hm, I had already run pip install black and fixed one error in test_dialcrowd, but I'm not seeing an error when I run ./autoformat.sh right now.

**Comment by EricMichaelSmith** [comment]
> Hm, I had already run pip install black and fixed one error in test_dialcrowd, but I'm not seeing an error when I run ./autoformat.sh right now. No files are changing (186 were left unchanged).

Hmm grasping at straws a bit - what if you add a second empty line above `except ImportError:` in that script? I notice that there are two above that line in `tests/crowdsourcing/tasks/turn_annotations_static/test_turn_annotations_static.py`

**Comment by JackUrb** [comment]
The crowdsourcing test failure is strange - it's failing to create the agents. Perhaps you need to manually import your specific blueprint type and ensure that the config you're feeding to the test includes this as well?

Also does this test pass locally? I'd expect you may find clearer debug output in your own console if not.

**Comment by jessicah25** [comment]
It looks like there's an issue with Mephisto/flask/black. Mephisto 1.0.1 and flask 1.1.4 require click <8.0, but it seems like black 22.3.0 requires click >=8.0 which runs into a compatibility issue.

**Comment by EricMichaelSmith** [comment]
> It looks like there's an issue with Mephisto/flask/black. Mephisto 1.0.1 and flask 1.1.4 require click <8.0, but it seems like black 22.3.0 requires click >=8.0 which runs into a compatibility issue. Otherwise, pytest is now passing on my end. I downgraded black to 20.8b1, then installed click 7.1.2, but it reformatted many other files, which may not be the desired outcome.

Yeah, not surprised - sometimes the ParlAI and Mephisto versions conflict, which leads to my personal recommendation (@JackUrb can weigh in) of having a separate env for each. Hmm did adding an extra line above `except ImportError:` fix the current failure in the lint CI check? Once that test is passing I'm personally happy to approve, because it looks like all currently working tests will have passed

**Comment by jessicah25** [comment]
Giving that a shot! When I autoformatted it, it took out the space again, so I'm manually adding that in now.

**Comment by EricMichaelSmith** [comment]
> Giving that a shot! When I autoformatted it, it took out the space again, so I'm manually adding that in now.

Ugh - okay, that's so weird that now the precommit CI check fails, because it wants to format the extra space out, but the lint one passes because it requires the extra space. @stephenroller have you ever seen an instance like this where these checks are in conflict with each other? @jessicah25 the only other obvious thing I can think of to say is that you can perhaps try pulling from `main` again, but I doubt that'll do anything because the PRs merged in since you did that don't obviously seem to be related to this

**Comment by jessicah25** [comment]
Seems like it solved the problem? I just rebased from main.

**Comment by EricMichaelSmith** [comment]
> Seems like it solved the problem? I just rebased from main.

Hmm I have to approve each run of the lint, etc. CI checks - looks like the problem still remains. Will try asking around to see if anyone else has seen this issue

**Comment by EricMichaelSmith** [comment]
> Seems like it solved the problem? I just rebased from main.

Hmm sorry can you re-commit a version where the second space after the class def is present? I wanna record what the pre-commit error is, since the log for that looks to be missing now for some reason. Then I can ask around with more complete information

**Comment by jessicah25** [comment]
Just did that!

**Comment by EricMichaelSmith** [comment]
> Just did that!

Great, thanks! Will ask about this now

**Comment by EricMichaelSmith** [comment]
> > Just did that!
> 
> Great, thanks! Will ask about this now

Okay cool, asked the others - will see what they say

**Comment by EricMichaelSmith** [comment]
> > > Just did that!
> > 
> > 
> > Great, thanks! Will ask about this now
> 
> Okay cool, asked the others - will see what they say

Just heard back - apparently black (i.e. the pre-commit CI check) should be the source of truth on this. Can you try adding an exception to .flake8 to suppress the lint error for this here?

**Comment by jessicah25** [comment]
I tried adding test_dialcrowd.py to the exclude files of .flake8, let me know if that was what you were referring to!

**Comment by EricMichaelSmith** [comment]
> I tried adding test_dialcrowd.py to the exclude files of .flake8, let me know if that was what you were referring to!

Ha, sorry - realize that wasn't clear =/ Hmm can you try looking up the lint code that corresponds to having 1 space and not 2, and then adding it to `extend-ignore` in that file? I think that should be a robust fix, because it'll prevent this issue from recurring

**Comment by EricMichaelSmith** [comment]
> This should be the error code we're looking for:
> 
> E305 | expected 2 blank lines after end of function or class
> 
> Hopefully this works! (Looks like I have to remove the space as well)

Ugh this is ridiculous - lint check still fails with the same issue:
```
[
  {
    path: 'tests/crowdsourcing/tasks/dialcrowd/test_dialcrowd.py',
    start_line: 76,
    end_line: 76,
    start_column: 1,
    end_column: 1,
    annotation_level: 'failure',
    message: '[BLK100] Black would make changes.'
  }
]
```
Did you say that, when you ran black on your side, it did or didn't flag having only one space? If it did flag it, I wonder if it's possible to re-run it on your end to check that the error code that you added really suppresses this error?

**Comment by jessicah25** [comment]
Just ran autoformat.sh; it keeps reformatting test_dialcrowd.py to the 2 spaces even with the E305 error suppressed.

**Comment by EricMichaelSmith** [comment]
> Just ran autoformat.sh; it keeps reformatting test_dialcrowd.py to the 2 spaces even with the E305 error suppressed. If I run autoformat.sh -f, which only runs flake8 (without only 1 space), the error doesn't seem to show up.

Okay, yeah, I've reproduced the error with a new test PR at https://github.com/facebookresearch/ParlAI/pull/4519/files, and I see the same error. I've tried various perturbations of this line to no avail, so I'm asking around to see if anyone else knows how to interpret that cryptic Black error

**Comment by jessicah25** [comment]
Sounds good!

**Comment by stephenroller** [comment]
Thanks for so much patience @jessicah25. Seems like we're only nits away. Let's just land this and I'll clean up black next week.

**Comment by stephenroller** [comment]
(I will request a `git merge main` to see if it resolves all the broken tests)

**Review by EricMichaelSmith — APPROVED** [review]
Approving - thanks for the patience on this @jessicah25 !

**Comment by EricMichaelSmith** [comment]
> (I will request a `git merge main` to see if it resolves all the broken tests)

@jessicah25 Hmm can you try doing a `git merge main` one last time to see how many tests are resolved before we merge this in? :)

**Comment by jessicah25** [comment]
I just ran git merge main in my dialcrowd branch; is that what you meant? It's all up to date.

**Comment by EricMichaelSmith** [comment]
> I just ran git merge main in my dialcrowd branch; is that what you meant? It's all up to date.

Ah, can you try merging `facebookresearch:main` into your fork once last time, as you did a few days ago? I'm not sure that I can do it myself on your fork - for instance, I tried following [this guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) but I can't see "Fetch upstream", implying that I may not have write access to it

**Comment by jessicah25** [comment]
just did that!

**Comment by EricMichaelSmith** [comment]
Can confirm that, as expected, the lint CI check is the only one that newly fails with this PR. Will merge now

**Comment by EricMichaelSmith** [comment]
Hi @jessicah25 ! I was wondering - do you have any interest in maintaining `dialcrowd` and keeping it up to date? @JackUrb suggested in https://github.com/facebookresearch/ParlAI/pull/4677 that `dialcrowd` may not work correctly currently because the frontend uses a pre-v1.0 version of Mephisto.

Otherwise, one option is for us to eventually remove the code itself from the HEAD of `main`, but leave the dialcrowd folder present, with a README giving instructions to the user for how to use dialcrowd by rewinding `main` to a known-working commit.

What do you think? Either way would work for us :)

CC @klshuster

**Comment by jessicah25** [comment]
Hi Eric! I would be willing to upkeep it for a bit; would I reopen another pull request as I'm doing so? I can look at the necessary updates next week.

**Comment by EricMichaelSmith** [comment]
> 

Hi @jessicah25 ! Yes, you could open another pull request to make the changes. Thanks!

**Pull request merged**
